### DISCUSSION TITLE
Move/Copy the active line OR selected text to another file

### DISCUSSION THREAD

**Discussion by GitMurf** [SilentVoid13/Templater#171] [Templates Showcase]
https://gist.github.com/GitMurf/fc4132acd805b9c1413ab84c6bd11576

Demo: https://user-images.githubusercontent.com/64155612/116498977-e575e380-a85f-11eb-9333-14b16cf18982.mp4

Move or Copy the active line (or selected text) of the active file to a chosen file.

Change the `const moveMode = "move";` to your choice. I recommend making multiple templates with the same code and then just changing this variable for the different "modes" outlined below.

Now can do the following:

- Move multi-lines if you select them before running script
- "Copy" instead of "move"
- Move but leave a block ref embed link behind in original location pointing to new location
- Leave in current location and create block ref and then choose where to copy a block ref embed link to in another file

**Comment by Quorafind** [comment]
Is it doable to add context by using dialog?

like I want to move `Youtube.com is a website` to another file.

Then I want to add text `START` above it, `This is an end` below it and then move 

```
START
Youtube.com is a website
This is an end
```

to another file.

OR 

Move text to another file and then insert text above it and below it? Depends on you

**Reply by GitMurf** [reply]
If you select multiple lines it will move them all. Is that sufficient? So you can add the START and then "this is the end" lines above and below, select all 3 lines, and then move it all.

**Comment by GitMurf** [comment]
Update v1.6.2 - Fixed removing of line on "move" as new Templater update changed the way it clears the line text.

https://gist.github.com/GitMurf/fc4132acd805b9c1413ab84c6bd11576

**Comment by chrisgrieser** [comment]
thanks for this script! it actually solves some problems for me :)

One minor issue that I noticed with the script: If you currently have text select (e.g. after highlighting a text passage with `== ==`) and then insert the template, the blockreference code `^123456` is inserted after the selected text and not at the end of the block, therefore breaking this script.

**Reply by chrisgrieser** [reply]
actually, the problem can be solved by removing the if-condition `if(tp.file.selection() == '')` and running the enclosed lines with or without selection.

**Comment by pcuellar** [comment]
Thanks a lot for this, it handles almost al scenarios I was thinking about. If I may suggest, one last mode that could complement this template will be to send a block reference to a target file instead of sending the block from current file to target file leaving a block reference in current file (as it is achieved by the `embed` mode in the script). It is similar to what you did [with this template](https://github.com/SilentVoid13/Templater/discussions/170) but instead of keeping the block reference in the clipboard, you have the option to send it to a target file.

**Reply by chrisgrieser** [reply]
I think you can already achieve that by setting the `moveMode` to `blockRef` in this script, see line 13-14 :)

**Reply by pcuellar** [reply]
Ohh shoot, somehow I missed this mode. Thanks @chrisgrieser for filling my eyes' gap 👀

**Comment by fedeescalada** [comment]
Great job! Just what I was looking for, thank you! :)

It would be great to have an option to force the target note to today's note. So you don't have to choose it from the GUI. I use to "bring" past meeting minutes to today's notes.

What do you think?

**Comment by EvoAlg** [comment]
The last post was a while ago.

I had an out of bounds error with the first instance of the line
`cmEditorAct.setSelection({ line: curLine, ch: 0 }, { line: curLine, ch: 9999 });`

so I changed it to

```
let lineContent = cmEditorAct.getLine(curLine);
let lineLength = lineContent.length;
cmEditorAct.setSelection({ line: curLine, ch: 0 }, { line: curLine, ch: lineLength });
```

**Comment by tom-rm-meyer-ISST** [comment]
Hey just went over the code as I currently want to support a workflow to extract to a new node including a template (see [discussion](https://github.com/SilentVoid13/Templater/discussions/1410)). This templates supports me in understanding how things are meant to be!

# Increase reusability 

>Change the const moveMode = "move"; to your choice. I recommend making multiple templates with the same code and then just changing this variable for the different "modes" outlined below.

I recommend you to do it differently for better maintainability. Maybe this is also a great update for [your tutorial on how to use templater](https://github.com/SilentVoid13/Templater/discussions/187).

1. Create folder structure similar to the following:
```
vault
- templates # folder for your markdown templates and everything to trigger
  - MOVE template
  - EMBED template
  - ... other combinations as wanted
- scripts # folder for reusable .js folders
  - move_file.js
```

2. Update your Templater settings accordingly (control + comma > Community Plugins > Templater)
- template folder location: templates
- script files folder location: scripts

3. Make the gist json reusable:
``` js
/** usage:
  *
  * tp: templater object to use
  * firstOrLastLine: 'first' will add to top of file. 'last' will add to bottom of file
  * bChooseLine: choose a specific line to move to underneath; overrules firstOrLastLine if true
  * bOpenFile after moving the line, open the file it was moved to
  * mobeMode: /"move" = default, "copy" = duplicate, "embed" = move and leave a block ref link, "blockRef" = create block ref and add **/
async function move_file (tp, firstOrLastLine, bChooseLine, bOpenFile, moveMode) {
  // your content without the header to steer functionality
}

module.exports = move_file;
```

4. Create the templates using parameters only (e.g. move-first-line)
``` js
<%* await tp.user.move_file(tp, 'first', false, false, 'move') %>
```

Edit: Checked how this works for my own template. Following notes:
- you need to pass the `tp` (templater object) to use the prompt features (see [docs](https://silentvoid13.github.io/Templater/user-functions/script-user-functions.html#global-namespace))
- you need to return a value that represents the `tR` (see [someone's blog](https://shbgm.ca/blog/obsidian/how-to-use-templater-js-scripts#The+JS+File))
- you need to access the global variable `app` without this reference (`this.app` -> `app`)
